### ISSUE TITLE
Regularized tensor decomposition

### ISSUE THREAD

**Issue by DylanMannKrzisnik** [tensorly/tensorly#169]
Dear all,

Apologies if this concern has already been dealt with. I was wondering wether Tensorly supports regularization of tensor decompositions. For instance, applying L1 regularization to obtain sparse PARAFAC factors.
Much of what I've seen regards decomposition of sparse tensors, rather than sparse decompositions of dense tensors.

As a reference: Cichocki A., Zdunek R. (2007) Regularized Alternating Least Squares Algorithms for Non-negative Matrix/Tensor Factorization. In: Liu D., Fei S., Hou Z., Zhang H., Sun C. (eds) Advances in Neural Networks – ISNN 2007. ISNN 2007. Lecture Notes in Computer Science, vol 4493. Springer, Berlin, Heidelberg

Thank you

**Comment by 20zinnm** [comment]
I'm also interested in this. How hard would it be to incorporate?

**Comment by JeanKossaifi** [comment]
We currently don't have this, would be great to add if you are interested in contributing!
The closest we currently have is the `sparsity` argument in `parafac`, which allows to express the tensor as a sum of a low-rank and a sparse tensor.

**Comment by DylanMannKrzisnik** [comment]
One way to go for PARAFAC L1-regularisation would be to apply the soft-thresholding operator of Eq. 6 from Hervas et al. 2019 (https://doi.org/10.1016/j.chemolab.2019.01.004) to the regression coefficients which result from alternating least-squares.

Operator:
B_lasso = sgn(B_LS)*(abs(B_LS) - lambda)

In other words, apply this operator each time ALS performs a regression to sparsify the final PARAFAC factor matrices. Maybe this method comes with pitfalls which I'm not aware of.

**Comment by JeanKossaifi** [comment]
@DylanMannKrzisnik that would be great! 

We already have proximal operators, including soft and SVD thresholding: https://github.com/tensorly/tensorly/blob/f72be606027f136c35da7c2d3e970e20d1f2a1ef/tensorly/tenalg/proximal.py#L9

For usage examples, you can check the [robust tensor PCA](https://github.com/tensorly/tensorly/blob/master/tensorly/decomposition/robust_decomposition.py) that uses them.

**Comment by DylanMannKrzisnik** [comment]
Great, I'll look into this.

**Comment by DylanMannKrzisnik** [comment]
Hello @JeanKossaifi,

I've made some (minor) changes needed to apply soft-thresholding to ALS. Before creating a pull request, I've been thinking about a script to test this implementation, as seen in the _contrib_ directory, in line with the contributing guidelines. I thought a comparison of PARAFAC factors with and without soft-thresholding for a given data-set could do the trick?

**Comment by DylanMannKrzisnik** [comment]
Quick update: performing soft-thresholding during ALS does not seem to be working quite well. For now, maybe a very simple alternative is to apply soft-thresholding on the estimated factors. In other words, apply soft-thresholding on the output rather than applying it during optimisation.

I'll share here whatever is found to work better.

**Comment by cohenjer** [comment]
So, supposing you are using an alternating procedure such as ALS, and supposing you are regularizing with the l1 norm of some factor (lets suppose you are regularizing only one factor to avoid theoretical caveats). What you are solving is in fact something like:
$ argmin_A \|T - A(B\odot C)^T \|_F^2 + \lambda \|A\|_1 $
which, while being a subproblem of the alternating algorithm, admits no closed-form solution (and is in fact a problem known as the [LASSO](https://en.wikipedia.org/wiki/Lasso_(statistics))). LASSO can be solved up to arbitrary precision but it typically takes an iterative algorithm to solve it, and I don't know wether we could make this efficient.

A way out, employed in the litterature, is to forget about ALS, and instead use alternating proximal gradient. In that case, you need to apply the proximal operator of the l1 norm, aka the soft thresholding operator you mentioned, on the update of the gradient.

**Comment by DylanMannKrzisnik** [comment]
Thanks for the input @cohenjer 

As you've underline, avoiding the computation of a penalized error function in an iterative framework would be desirable. I'd like to look into this alternating proximal gradient approach you mention here. Would you know of any good references for a first dive into this technique?

**Comment by cohenjer** [comment]
Hey,

A quite well-cited paper that proposes such an algorithm is "A BLOCK COORDINATE DESCENT METHOD FOR REGULARIZED MULTI-CONVEX OPTIMIZATION WITH APPLICATIONS TO NONNEGATIVE TENSOR FACTORIZATION AND COMPLETION" available on arxiv. I think one can easily make a faster proximal gradient-type algorithm (and many have already done that) but its rather straightforward and well-understood.

**Comment by DylanMannKrzisnik** [comment]
Thank you, looking forward to reading this

**Comment by cohenjer** [comment]
@DylanMannKrzisnik In principle, once #284, you should be able to run sparse Parafac. The algorithm used to impose sparsity is Alternating Descent Method of Multipliers, the relevant pseudo-code and explanations are found is a paper by K. Huang and coauthors [(link)](https://arxiv.org/abs/1506.04209).

If you want nonnegativity as well, in #224 we implemented nonnegative CP using Alternating nonnegative least squares, and l1 sparsity is a built-in feature.

I think once #284 is merged we can close this issue ? @JeanKossaifi ?

**Comment by JeanKossaifi** [comment]
Yes I agree. Let's decide on the API https://github.com/tensorly/lab/pull/1 and merge #284

**Comment by JeanKossaifi** [comment]
This has been added, closing.
